File: language/src/main/java/language/nodes/expr/ReadArgumentNode.java
package language.nodes.expr;

import com.oracle.truffle.api.frame.VirtualFrame;
import com.oracle.truffle.api.profiles.BranchProfile;
import language.runtime.CrNull;
import org.jetbrains.annotations.NotNull;

public class ReadArgumentNode extends ExprNode {
    private final int index;

    private final BranchProfile outOfBoundsTaken = BranchProfile.create();

    public ReadArgumentNode(int index) {
        this.index = index;
    }

    @NotNull
    @Override
    public Object executeGeneric(VirtualFrame frame) {
        Object[] args = frame.getArguments();
        if (index < args.length) {
            return args[index];
        } else {
            outOfBoundsTaken.enter();
            return CrNull.INSTANCE;
        }
    }
}

File: language/src/main/java/language/nodes/expr/ReadArgumentNode.kt
package language.nodes.expr

import com.oracle.truffle.api.frame.VirtualFrame
import com.oracle.truffle.api.profiles.BranchProfile
import language.runtime.CrNull

class ReadArgumentNode(private val index: Int) : ExprNode() {
    private val outOfBoundsTaken = BranchProfile.create()
    override fun executeGeneric(frame: VirtualFrame): Any {
        val args = frame.arguments
        return if (index < args.size) {
            args[index]
        } else {
            outOfBoundsTaken.enter()
            CrNull
        }
    }
}